conda/gateways/repodata/__init__.py

```diff
@@ -289,3 +290,97 @@ def conda_http_errors(url, repodata_fn):
             e.response,
             caused_by=e,
         )
+
+
+class CacheJsonState(dict):
+    """
+    Load/save `.state.json` that accompanies cached `repodata.json`
+    """
+
+    def __init__(self, cache_path_json, cache_path_state, repodata_fn):
+        self.cache_path_json = pathlib.Path(cache_path_json)
+        self.cache_path_state = pathlib.Path(cache_path_state)
+        self.repodata_fn = repodata_fn
+
+    def load(self):
+        """
+        Cache headers and additional data needed to keep track of the cache are
+        stored separately, instead of the previous "added to repodata.json"
+        arrangement.
+        """
+        try:
+            state_path = self.cache_path_state
+            log.debug("Load %s cache from %s", self.repodata_fn, state_path)
+            with state_path.open("r") as s:
+                state = json.load(s)
+            # json and state files should match
+            json_stat = self.cache_path_json.stat()
+            if not (
+                state.get("mtime_ns") == json_stat.st_mtime_ns
+                and state.get("size") == json_stat.st_size
+            ):
+                # clear mod, etag, cache_control to encourage re-download
+                state.update({"etag": "", "mod": "", "cache_control": "", "size": 0})
+            for field in (
+                "etag",
+                "mod",
+                "cache_control",
+                "size",
+            ):
+                alias = field
+                if alias in state:
+                    self[field] = state[alias]
```

Comment:
We should define a private class level `_fields` variable (e.g., `_fields = ("etag", "mod", "cache_control", "size")`).

```suggestion
            self.update({field: state[field] for field in self._fields if field in state})
```